Bottom: 70ec768bb432e2c8f09f5f338afbe7f79e73f25d
Top:    ce6ed6f8984b691eeae2dd1740b5e403fe88489c
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-25 19:44:55 +0100

draw.c: Split "Gathering" from real drawing routines (Element names)


---

diff --git a/src/draw.c b/src/draw.c
index 80b7fda..168884a 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -196,6 +196,24 @@ element_callback (const BoxType * b, void *cl)
   return 1;
 }
 
+static void
+draw_element_name (ElementType *element)
+{
+  if (gui->gui && TEST_FLAG (HIDENAMESFLAG, PCB))
+    return;
+  if (TEST_FLAG (HIDENAMEFLAG, element))
+    return;
+  if (doing_pinout || doing_assy)
+    gui->set_color (Output.fgGC, PCB->ElementColor);
+  else if (TEST_FLAG (SELECTEDFLAG, &ELEMENT_TEXT (PCB, element)))
+    gui->set_color (Output.fgGC, PCB->ElementSelectedColor);
+  else if (FRONT (element))
+    gui->set_color (Output.fgGC, PCB->ElementColor);
+  else
+    gui->set_color (Output.fgGC, PCB->InvisibleObjectsColor);
+  DrawTextLowLevel (&ELEMENT_TEXT (PCB, element), PCB->minSlk);
+}
+
 static int
 name_callback (const BoxType * b, void *cl)
 {
@@ -207,7 +225,7 @@ name_callback (const BoxType * b, void *cl)
     return 0;
 
   if (ON_SIDE (element, *side))
-    DrawElementName (element);
+    draw_element_name (element);
   return 0;
 }
 
@@ -1411,7 +1429,10 @@ void
 DrawElement (ElementTypePtr Element)
 {
   DrawElementPackage (Element);
-  DrawElementName (Element);
+  if (Gathering)
+    DrawElementName (Element);
+  else
+    draw_element_name (Element);
   DrawElementPinsAndPads (Element);
 }
 
@@ -1421,19 +1442,11 @@ DrawElement (ElementTypePtr Element)
 void
 DrawElementName (ElementTypePtr Element)
 {
-  if (gui->gui && TEST_FLAG (HIDENAMESFLAG, PCB))
-    return;
+  assert (Gathering);
+
   if (TEST_FLAG (HIDENAMEFLAG, Element))
     return;
-  if (doing_pinout || doing_assy)
-    gui->set_color (Output.fgGC, PCB->ElementColor);
-  else if (TEST_FLAG (SELECTEDFLAG, &ELEMENT_TEXT (PCB, Element)))
-    gui->set_color (Output.fgGC, PCB->ElementSelectedColor);
-  else if (FRONT (Element))
-    gui->set_color (Output.fgGC, PCB->ElementColor);
-  else
-    gui->set_color (Output.fgGC, PCB->InvisibleObjectsColor);
-  DrawTextLowLevel (&ELEMENT_TEXT (PCB, Element), PCB->minSlk);
+  AddPart (&ELEMENT_TEXT (PCB, Element));
 }
 
 /* ---------------------------------------------------------------------------
@@ -1707,9 +1720,11 @@ EraseElementPinsAndPads (ElementTypePtr Element)
 void
 EraseElementName (ElementTypePtr Element)
 {
+  assert (Gathering);
+
   if (TEST_FLAG (HIDENAMEFLAG, Element))
     return;
-  DrawTextLowLevel (&ELEMENT_TEXT (PCB, Element), PCB->minSlk);
+  AddPart (&ELEMENT_TEXT (PCB, Element));
 }
